### PR TITLE
[RISCV] Add isel special case for (and (shl X, c2), c1) -> (slli_uw (srli x, c4-c2), c4).

### PR DESCRIPTION
**Pull request by topperc**
Where c1 is a shifted mask with 32 set bits and c4 trailing zeros.

This is an alternative to #91626.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1350,6 +1350,21 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
             return;
           }
         }
+
+        // Turn (and (shl x, c2), c1) -> (slli_uw (srli x, c4-c2), c4) where c1
+        // is shifted mask with 32 set bits and c4 trailing zeros.
+        unsigned Trailing = llvm::countr_zero(C1);
+        if (Leading + Trailing == 32 && C2 < Trailing &&
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:1357]: Do we need a !IsCANDI check here?

**Review by preames — APPROVED** [review]
LGTM w/minor comment

**Comment by dtcxzyw** [comment]
Emm, https://github.com/llvm/llvm-project/pull/91626 seems to perform better than this patch. BTW this patch converts some `slli` insts into `slli.uw`, which makes them less compressible.

**Comment by topperc** [comment]
> Emm, https://github.com/llvm/llvm-project/pull/91626 seems to perform better than this patch. BTW this patch converts some `slli` insts into `slli.uw`, which makes them less compressible.
> 
> 

Can you get reproducers?

**Comment by dtcxzyw** [comment]
> > Emm, #91626 seems to perform better than this patch. BTW this patch converts some `slli` insts into `slli.uw`, which makes them less compressible.
> 
> Can you get reproducers?

godbolt: https://godbolt.org/z/sMv9fzWbY
```
; llc -mtriple=riscv64 -mattr=+c,+m,+zba -o -
define i64 @func0000000000000000(i64 %0) #0 {
entry:
  %1 = mul i64 %0, 100
  %2 = udiv i64 %1, 70
  %3 = shl i64 %2, 32
  ret i64 %3
}
```

Before:
```
func0000000000000000:                   # @func0000000000000000
        lui     a1, %hi(.LCPI0_0)
        ld      a1, %lo(.LCPI0_0)(a1)
        li      a2, 100
        mul     a0, a0, a2
        mulhu   a0, a0, a1
        srli    a0, a0, 6
        slli    a0, a0, 32
        ret
```

After:
```
func0000000000000000:                   # @func0000000000000000
	lui	a1, %hi(.LCPI1_0)
	ld	a1, %lo(.LCPI1_0)(a1)
	li	a2, 100
 	mul	a0, a0, a2
 	mulhu	a0, a0, a1
 	srli	a0, a0, 6
 	slli.uw	a0, a0, 32 *!!!*
 	ret
```

**Comment by topperc** [comment]
Need a check that Leading is non-zero or Trailing is less than 32

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by topperc** [comment]
> LGTM.

Are the results better now? Was that the only issue?

I'll add a test for that case before I commit.

**Comment by dtcxzyw** [comment]
> > LGTM.
> 
> Are the results better now? Was that the only issue?
> 

Yeah, I confirmed that all regressions have been fixed :)
See https://github.com/dtcxzyw/llvm-codegen-benchmark/commit/5c1d00242e58f98cce06ff7e35137209e8f45bff

**Pull request merged**